node/src/tasks/msp_move_bucket.rs

```diff
@@ -140,175 +149,421 @@ where
+                            error
                         );
                         continue;
                     }
+                };
+
+                if proven.len() != 1 {
+                    error!(
+                        target: LOG_TARGET,
+                        "Expected exactly one proven chunk but got {}",
+                        proven.len()
+                    );
+                    continue;
+                }
 
-                    let chunk_data = proven[0].data.clone();
-                    let chunk_id = ChunkId::new(chunk);
+                let chunk_data = proven[0].data.clone();
+                let chunk_id = ChunkId::new(chunk);
 
-                    if chunk_id != proven[0].key {
-                        error!(
-                            target: LOG_TARGET,
-                            "Expected chunk id {:?} but got {:?}",
-                            chunk, proven[0].key
-                        );
-                        continue;
-                    }
+                if chunk_id != proven[0].key {
+                    error!(
+                        target: LOG_TARGET,
+                        "Expected chunk id {:?} but got {:?}",
+                        chunk, proven[0].key
+                    );
+                    continue;
+                }
 
-                    if let Err(error) = self
-                        .storage_hub_handler
-                        .file_storage
-                        .write()
-                        .await
-                        .write_chunk(&file_key, &chunk_id, &chunk_data)
-                    {
-                        // TODO: Handle this error better.
-                        error!(
-                            target: LOG_TARGET,
-                            "Failed to write chunk {:?} of file {:?} to storage: {:?}",
-                            chunk, file_key, error
-                        );
-                    } else {
-                        // We successfully downloaded the chunk, so we can break out of the retry loop.
-                        break;
-                    }
+                if let Err(error) = self
+                    .storage_hub_handler
+                    .file_storage
+                    .write()
+                    .await
+                    .write_chunk(&file_key, &chunk_id, &chunk_data)
+                {
+                    error!(
+                        target: LOG_TARGET,
+                        "Failed to write chunk: {:?}",
+                        error
+                    );
+                } else {
+                    downloaded = true;
+                    break;
                 }
             }
 
+            if !downloaded {
+                error!(
+                    target: LOG_TARGET,
+                    "Failed to download chunk {:?} after {} retries",
+                    chunk,
+                    DOWNLOAD_REQUEST_RETRY_COUNT
+                );
+            }
```

Comment:
I assume this was intentional to attempt to download all files and ignore files that have not successfully been downloaded.

But we should add a TODO to have some method to either track missing files for a bucket so that an MSP may resolve this manually afterwards or implement some automatic re-run procedure that could be started as an RPC to re-attempt downloading any missing ones.